src/platforms/common/blackpill-f4/usbdfu.c

```diff
@@ -75,4 +80,46 @@ int main(void)
 
 void dfu_event(void)
 {
+	static bool idle_state = false;
+	/* Ask systick to pause blinking for 1 second */
+	dfu_activity_counter = 10;
+	/* Blink it ourself */
+	SET_IDLE_STATE(idle_state);
+	idle_state = !idle_state;
+}
+
+static void sys_tick_init(void)
+{
+	/* Use SysTick at 10Hz to toggle-blink the blue LED at 5 Hz */
+	systick_set_clocksource(STK_CSR_CLKSOURCE_AHB_DIV8);
+	systick_set_reload(rcc_ahb_frequency / 8 / 10);
+	/* SYSTICK_IRQ with low priority */
+	nvic_set_priority(NVIC_SYSTICK_IRQ, 14U << 4U);
+	systick_interrupt_enable();
+	/* Start the heartbeat timer */
+	systick_counter_enable();
+}
+
+void sys_tick_handler(void)
+{
+	static int count = 0;
+	if (dfu_activity_counter > 0) {
+		dfu_activity_counter--;
+		return;
+	}
+	switch (count) {
+	case 0:
+		/* Reload downcounter */
+		count = 10;
+		SET_IDLE_STATE(false);
+		break;
+	case 1:
+		count--;
```

Comment:
If you wanted to, you could eliminate this line and the break on this case, putting in `/* fallthrough */` in place of the break, wonder if that would decrease the footprint of this slightly?

Reply:
This is a mentally taxing transformation, I just wanted to express intent (of changing the counter to next step) and let the compiler deal with what's necessary and possible. Thinking harder I see how this could work:
```c
	case 1:
		/* Blink like a very slow PWM */
		SET_IDLE_STATE(true);
		/* fallthrough */
	default:
		count--;
		break;
	}
```
Again, the first sector of F4 is 16 KiB, and the bootloader is currently slightly above 8 KiB (or slightly below when resorting to newlib-nano) because of all of the libopencm3 usb code etc. Space is not an issue here. I'd rather see (or at least test) BOS descriptors for DFU added, if they are available (and how much space do they take up).

Reply:
Yeah, that's the transformation we were thinking of, but that's a fair point about additional mental burden - in that case, don't do it. This was very much a suggestion to talk about it, not something that has to be done.

Good questions about the MS OS Descriptors and BOS descriptor - we'd guess it should cause around 100 to 150 bytes of usage, but that'd be fascinating to see. Guessing you're planning on poking that in a follow-up PR?

Reply:
No, I am not that well versed in USB descriptors, especially MS extensions. It would be nice to try to add them so that WINUSB gets assigned automatically to new builds of BMPBootloader, and compare that with the bmputil/libwdi workflow/UX. There are 150 lines of descriptor structs and macros in `usb_descriptors.h` for the payload part, including a WINUSB-compatible DFU runtime interface description.

Reply:
Ah, fair - no worries. Well, we can look at poking it and we can run you through how it works but that'll be in a couple of days at the current rate of things. Agreed that it would be nice!